Sources/Rendering/WebGPU/UniformBuffer/index.js

```diff
@@ -0,0 +1,338 @@
+          },
+        ],
+      });
+      model.sendTime.modified();
+      model.sendDirty = false;
+    }
+
+    // send data down if needed
+    if (model.sendDirty) {
+      device
+        .getHandle()
+        .queue.writeBuffer(
+          model.UBO.getHandle(),
+          0,
+          model.arrayBuffer,
+          0,
+          model.sizeInBytes
+        );
+      model.sendTime.modified();
+      model.sendDirty = false;
+    }
+  };
+
+  publicAPI.createView = (type) => {
+    if (type in model === false) {
+      if (!model.arrayBuffer) {
+        model.arrayBuffer = new ArrayBuffer(model.sizeInBytes);
+      }
+      model[type] = new window[type](model.arrayBuffer);
+    }
+  };
+
+  publicAPI.setValue = (name, val) => {
+    publicAPI.sortBufferEntries();
+    const idx = model.bufferEntryNames.get(name);
+    if (idx === undefined) {
+      vtkErrorMacro(`entry named ${name} not found in UBO`);
+      return;
+    }
+    const entry = model.bufferEntries[idx];
+    publicAPI.createView(entry.nativeType);
+    const view = model[entry.nativeType];
+    if (view[entry.offset / view.BYTES_PER_ELEMENT] !== val) {
+      view[entry.offset / view.BYTES_PER_ELEMENT] = val;
+      model.sendDirty = true;
+    }
+  };
+
+  publicAPI.setArray = (name, arr) => {
+    publicAPI.sortBufferEntries();
+    const idx = model.bufferEntryNames.get(name);
+    if (idx === undefined) {
+      vtkErrorMacro(`entry named ${name} not found in UBO`);
+      return;
+    }
+    const entry = model.bufferEntries[idx];
+    publicAPI.createView(entry.nativeType);
+    const view = model[entry.nativeType];
+    for (let i = 0; i < arr.length; i++) {
+      if (view[entry.offset / view.BYTES_PER_ELEMENT + i] !== arr[i]) {
+        view[entry.offset / view.BYTES_PER_ELEMENT + i] = arr[i];
+        model.sendDirty = true;
+      }
+    }
+  };
+
+  publicAPI.getNativeView = (name) => {
+    publicAPI.sortBufferEntries();
+    const idx = model.bufferEntryNames.get(name);
+    if (idx === undefined) {
+      vtkErrorMacro(`entry named ${name} not found in UBO`);
+      return undefined;
+    }
+    const entry = model.bufferEntries[idx];
+    publicAPI.createView(entry.nativeType);
+    const bpe = model[entry.nativeType].BYTES_PER_ELEMENT;
+    model.sendDirty = true;
+    return new window[entry.nativeType](
+      model.arrayBuffer,
+      entry.offset,
+      entry.sizeInBytes / bpe
+    );
+  };
+
+  publicAPI.getSendTime = () => model.sendTime.getMTime();
+  publicAPI.getShaderCode = () => {
+    // sort the entries
+    publicAPI.sortBufferEntries();
+
+    let result = `[[block]] struct ${model.name}Struct\n{\n`;
+    for (let i = 0; i < model.bufferEntries.length; i++) {
+      const entry = model.bufferEntries[i];
+      result += `[[offset(${entry.offset})]] ${entry.name}: ${entry.type};\n`;
+    }
+    result += `};\n[[binding(${model.binding}), group(${model.group})]] var<uniform> ${model.name}: ${model.name}Struct;`;
+    return result;
+  };
```

Comment:
It is usually better / faster to use an array and just join it with `\n` at the end than doing a `+=` on string.

```
const lines = [
   `[[block]] struct ${model.name}Struct`,
   '{',
    model.bufferEntries.map(({offset, name, type}) => `[[offset(${offset})]] ${name}: ${type}`).join('\n'),
   '};',
   `[[binding(${model.binding}), group(${model.group})]] var<uniform> ${model.name}: ${model.name}Struct;`
];
return lines.join('\n');
```